src/main/host/descriptor/epoll.c

```diff
@@ -646,6 +649,18 @@ gint epoll_getEvents(Epoll* epoll, struct epoll_event* eventArray, gint eventArr
 
     trace("epoll descriptor %i collected %i events", descriptor_getHandle(&epoll->super), eventIndex);
 
+    GHashTableIter iter;
+    g_hash_table_iter_init(&iter, epoll->ready);
+    gpointer val;
+
+    /* We modified some watched objects above, so remove any that are no longer ready. */
+    while (g_hash_table_iter_next(&iter, NULL, (gpointer)&val)) {
+        EpollWatch* watch = val;
+        if (!_epollwatch_isReady(watch)) {
+            g_hash_table_iter_remove(&iter);
+        }
+    }
+
```

Comment:
I don't think we want to go through ALL of the watches when we only possibly changed a small number of them in the while loop above.

Instead of iterating the hash table here, how about we add the `isReady` check above immediately following line 634 (but still in the `if(_epollwatch_isReady(watch)) {` scope? That way we are only checking those events that we edit in the while loop, for which we may have caused the ready status to change (with the edits on line 629 and 633). If we have caused the event to become not ready, we could add it to a list or queue, and replace this hash table iterator with the list or queue iterator that removes those watches from the ready table.

Does this make sense?

Reply:
I changed it to store them in a list, and then later remove those from the hash table. Two downsides are (1) it requires memory allocations for the list, and (2) instead of looping over the hash table items, we need to iterate over the keys and perform hash table lookups for the items. But I don't think these downsides are very important.